grass7/raster/r.prominence/main.c

```diff
@@ -0,0 +1,526 @@
+
+    flag.absolute = G_define_flag();
+    flag.absolute->key = 'a';
+    flag.absolute->description = "Calculate absolute differences";
+
+    flag.localnorm = G_define_flag();
+    flag.localnorm->key = 'l';
+    flag.localnorm->description = "Local data normalisation";
+
+    flag.globalnorm = G_define_flag();
+    flag.globalnorm->key = 'g';
+    flag.globalnorm->description = "Global data normalisation";
+
+    flag.rect = G_define_flag();
+    flag.rect->key = 'r';
+    flag.rect->description = "Use quadratic neighbourhood";
+
+    flag.quiet = G_define_flag();
+    flag.quiet->key = 'q';
+    flag.quiet->description = "Disable on-screen progress display";
+
+
+    /* parse command line */
+    if (G_parser(argc, argv))
+        exit(EXIT_FAILURE);
+
+    G_get_window(&region);
+    nrows = Rast_window_rows();
+    ncols = Rast_window_cols();
+
+    /* check parameters for validity */
+    radius = atoi(parm.radius->answer);
+    if ((radius < 1) || (radius > nrows / 2) || (radius > ncols / 2)) {
+        G_fatal_error
+            ("Radius in rows/columns must be > 1 and smaller than half the map's extent in both directions.");
+    }
+    if ((flag.localnorm->answer) && (flag.globalnorm->answer)) {
+        G_fatal_error("Please choose either local OR global normalisation.");
+    }
+    mapset = G_calloc(512, sizeof(char));
+    mapset = G_find_raster(parm.input->answer, "");
+    if (mapset == NULL) {
+        G_fatal_error("Input map does not exist in the current location.");
+    }
+    if (!Rast_map_is_fp(parm.input->answer, mapset)) {
+        G_fatal_error
+            ("Input map is not a floating point map.\nOnly floating point maps are allowed as input maps.");
+    }
+    fd = Rast_open_old(parm.input->answer, mapset);
+    if (fd < 0) {
+        G_fatal_error("Could not open input map for reading!\n");
+    }
+    if (!G_legal_filename(parm.output->answer)) {
+        G_fatal_error("%s is not a legal filename for an output map.\n",
+                      parm.output->answer);
+    }
+    fd_out = Rast_open_new(parm.output->answer, DCELL_TYPE);
+    if (fd_out < 0) {
+        G_fatal_error("Could not open output map for writing!\n");
+    }
+
+    diskrow = Rast_allocate_d_buf();
+    outrow = Rast_allocate_d_buf();
+
+    /* initialize statistics */
+    n = 0;
+    firstrun = 1;
+    min = 0;
+    max = 0;
+    sum = 0.0;
+
+    if (!flag.quiet->answer) {
+        fprintf(stdout, "Calculating prominence index:\n");
+        fflush(stdout);
+    }
+
+    /* main loop: read raster row from disk, add to values in neighbourhood */
+    i = 0;
+    j = 0;
+    nullvalue = 0;
+    for (y = 0; y < nrows; y++) {
+        for (x = 0; x < ncols; x++) {
+            Rast_get_d_row(fd, diskrow, y);
+            centerval = diskrow[x];
+            if (Rast_is_d_null_value(&diskrow[x])) {
+                nullvalue = 1;
+            }
+            else {
+                nullvalue = 0;
```

Comment:
```suggestion
                nullvalue = 0;
                sum = 0;
                n = 0;
```
@benducke I guess sum and n need to be reset for each neighborhood/moving window, otherwise sum and n accumulate over the whole raster? Equivalent for the localnorm and globalnorm loops below?